File: src/main/java/ru/org/linux/user/UserLogPrepareService.java
/*
 * Copyright 1998-2023 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.user;

import com.google.common.collect.ImmutableMap;
import org.springframework.stereotype.Service;
import ru.org.linux.spring.dao.UserAgentDao;

import javax.annotation.Nonnull;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static ru.org.linux.util.StringUtil.escapeHtml;

@Service
public class UserLogPrepareService {
  private final UserService userService;
  private final UserAgentDao userAgentDao;

  private final static ImmutableMap<String, String> OPTION_DESCRIPTION;

  static {
    ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();

    builder.put(UserLogDao.OPTION_BONUS, "Изменение score");
    builder.put(UserLogDao.OPTION_NEW_EMAIL, "Новый email");
    builder.put(UserLogDao.OPTION_NEW_USERPIC, "Новая фотография");
    builder.put(UserLogDao.OPTION_OLD_EMAIL, "Старый email");
    builder.put(UserLogDao.OPTION_OLD_INFO, "Старый текст информации");
    builder.put(UserLogDao.OPTION_OLD_USERPIC, "Старая фотография");
    builder.put(UserLogDao.OPTION_REASON, "Причина");

    OPTION_DESCRIPTION = builder.build();
  }

  public UserLogPrepareService(UserService userService, UserAgentDao userAgentDao) {
    this.userService = userService;
    this.userAgentDao = userAgentDao;
  }

  @Nonnull
  public List<PreparedUserLogItem> prepare(@Nonnull List<UserLogItem> items) {
    return items.stream().map(item -> {
      Map<String, String> options = new HashMap<>();

      for (Map.Entry<String, String> option : item.getOptions().entrySet()) {
        String key = OPTION_DESCRIPTION.get(option.getKey());
        if (key == null) {
          key = escapeHtml(option.getKey());
        }

        String value;

        switch (option.getKey()) {
          case UserLogDao.OPTION_OLD_USERPIC:
          case UserLogDao.OPTION_NEW_USERPIC:
            value = "<a href=\"/photos/" + escapeHtml(option.getValue()) + "\">" + escapeHtml(option.getValue()) + "</a>";
            break;
          case UserLogDao.OPTION_IP:
            value = "<a href=\"/sameip.jsp?ip=" + escapeHtml(option.getValue()) + "\">" + escapeHtml(option.getValue()) + "</a>";
            break;
          case UserLogDao.OPTION_INVITED_BY:
            User user = userService.getUserCached(Integer.parseInt(option.getValue()));

            value = "<a href=\"/people/" + user.getNick() + "/profile\">" + user.getNick() + "</a>";

            break;
          case UserLogDao.OPTION_USET_AGENT:
            int id = Integer.parseInt(option.getValue());

            if (id != 0) {
              value = "<a href=\"/sameip.jsp?ua=" + id + "\">" + userAgentDao.getUserAgentById(id).orElse("<не найден>") + "</a>";
            } else {
              value = "<нет>";
            }

            break;
          default:
            value = escapeHtml(option.getValue());
            break;
        }

        options.put(key, value);
      }

      return new PreparedUserLogItem(item, userService.getUserCached(item.getActionUser()), options);
    }).collect(Collectors.toList());
  }
}

File: src/main/scala/ru/org/linux/user/UserLogPrepareService.scala
/*
 * Copyright 1998-2026 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package ru.org.linux.user

import org.joda.time.DateTimeZone
import org.springframework.stereotype.Service
import ru.org.linux.msgbase.UserAgentDao
import ru.org.linux.site.DateFormats
import ru.org.linux.user.UserLogDao.*
import ru.org.linux.util.StringUtil.escapeHtml

import java.time.Instant
import java.util.Date

object UserLogPrepareService {
  private val OptionDescription: Map[String, String] =
    Map(OptionBonus -> "Изменение score",
        OptionNewEmail -> "Новый email",
        OptionNewUserpic -> "Новая фотография",
        OptionOldEmail -> "Старый email",
        OptionOldInfo -> "Старый текст информации",
        OptionOldUserpic -> "Старая фотография",
        OptionReason -> "Причина",
        OptionUntil -> "Срок действия")
}

@Service
class UserLogPrepareService(userService: UserService, userAgentDao: UserAgentDao) {
  def prepare(items: collection.Seq[UserLogItem], timezone: DateTimeZone): Seq[PreparedUserLogItem] = {
    items.view.map((item: UserLogItem) => {
      val options = for ((rawKey, rawValue) <- item.options) yield {
        val key = UserLogPrepareService.OptionDescription.getOrElse(rawKey, escapeHtml(rawKey))

        val value = rawKey match {
          case OptionOldUserpic | OptionNewUserpic =>
            s"<a href=\"/photos/${escapeHtml(rawValue)}\">${escapeHtml(rawValue)}</a>"
          case OptionIp =>
            s"<a href=\"/sameip.jsp?ip=${escapeHtml(rawValue)}\">${escapeHtml(rawValue)}</a>"
          case OptionInvitedBy =>
            val user = userService.getUserCached(rawValue.toInt)
            s"<a href=\"/people/${user.nick}/profile\">${user.nick}</a>"
          case OptionUserAgent =>
            val id = rawValue.toInt
            val ip = item.options.getOrElse(OptionIp, "")

            if (id != 0) {
              s"<a href=\"/sameip.jsp?ua=$id&ip=$ip&mask=0\">${userAgentDao.getUserAgentById(id).orElse(escapeHtml("<не найден>"))}</a>"
            } else {
              escapeHtml("<нет>")
            }
          case OptionUntil =>
            val until = Instant.parse(rawValue)

            DateFormats.formatDefault(timezone, Date.from(until))
          case _ =>
            escapeHtml(rawValue)
        }

        key -> value
      }

      PreparedUserLogItem(item, userService.getUserCached(item.actionUser), options)
    }).toVector
  }
}
